File: MyCode/JavaScript/src/Competition.js
// create class called Competition
class Competition {
	constructor (newTitle, newYear) {
		this.title = newTitle
		this.year = newYear
		// futureproof lists for each Premiership and Championship
		// for crossgames.
		this.allPDTeams = []
		this.allCDTeams = []
		// whole lists of team for Premiership and Championship based onrank
		this.allTeams = []
		// there are number of games each week, and number of weeks
		this.allGames = [[],[],[],[],[],[],[],[],[]]
		// will be better way, but this will be used in getCrossOverGames.
		this.allPDRanks = []
		this.allCDRanks = []
	}

	// addTeam method, to add teams into each lists.
	addTeam(newRank, newName, newVenue, newCity){
		let newTeam = new Team(newRank, newName, newVenue, newCity)
		// To use for games home rank vs away rank
		this.allTeams.push(newTeam)
		// I know magic number is not a great way to do so, but for this assignment,
		// only have 14 teams, so used magic number
		// rank between 1~7 are in Premiership Division
		if (newRank <= 7){
			this.allPDTeams.push(newTeam)
			this.allPDRanks.push(newTeam.rank)
		}
		// rank between 8~14 are in Championship Division
		else {
			this.allCDTeams.push(newTeam)
			this.allCDRanks.push(newTeam.rank)
		}

	}

	// getTeam method, will print out all teams from each division.
	getTeams() {
		let result = `TEAMS ${View.NEWLINE()}`
		result += `Premiership Division ${View.NEWLINE()}`

		for (let aTeam of this.allPDTeams){
			result += aTeam + `${View.NEWLINE()}`
		}
		result += `Championship Division ${View.NEWLINE()}`
		for (let aTeam of this.allCDTeams){
			result += aTeam + `${View.NEWLINE()}`
		}
		return result
	}

	// addGame method, add all games into list.
	addGame(newWeek, newHomeTeamRank, newAwayTeamRank, newDateTime){
		let  newGame = new Game(newWeek,
														newHomeTeamRank,
														newAwayTeamRank,
														newDateTime)
		// as list's index start from 0 but there is no week 0. so minus 1 from week
		this.allGames[newWeek-1].push(newGame)
	}

	// getGames method, from list, printout on html to see the list of games.
	getGames() {
		let schedule = this.year + `${View.SPACE()}` + this.title + ` Draw ${View.NEWLINE()}`
		for (let aWeek of this.allGames){
			schedule += "Week:" + aWeek[0].week + `${View.NEWLINE()}`
			for (let aGame of aWeek){
				// to get the date type, e.g. Mon, Jul 16 2018
				let newDate = new Date(aGame.dateTime).toDateString()
				// adding that date
				schedule += newDate + `${View.SPACE()}`
				// getting time like 7:35PM
				// was gonna use navigator.language instead of en-US but it comes up with korean
				// even thought default language is ENG
				let newTime = new Date(aGame.dateTime).toLocaleTimeString("en-US", {hour: "numeric", minute: "numeric"})
				schedule += newTime + `${View.SPACE()}`

									// getting name for home team from rank
				schedule += "Team: " + this.allTeams[aGame.homeTeamRank - 1].name
									+ `${View.SPACE()}`
									+ "vs"
									+ `${View.SPACE()}`
									// getting name for away team from rank
									+ this.allTeams[aGame.awayTeamRank - 1].name
									+ `${View.SPACE()}`
									// getting venue for home team from rank
									+ "At: " + this.allTeams[aGame.homeTeamRank - 1].venue
									// getting city for home team from rank
									+ ", " + this.allTeams[aGame.homeTeamRank - 1].city
									+ `${View.NEWLINE()}`
			}
		}
		return schedule
	}

	// getCanterburyGames method, was thinking to call getGames as based on the teamName,
	// my method will print out different result but pdf asks specific name for this method
	// But still for futureproof, this method will receive variable called teamName
	// which is a string, that we want to get result out of.
	getCanterburyGames(teamName) {
		// find the rank of the team we want, as I dont want to use magic number. futureproof
		for (let aTeam of this.allTeams){
			if (aTeam["name"] == teamName){
				var aRank = aTeam["rank"]
			}
		}
		// again, fufure proofing print line, instead of fixed string with Canterbury
		let specificTeam = "Will only display " + teamName + " team games" + `${View.NEWLINE()}`
		// allGames will have multiple lists inside as there are more than one week of games
		for (let eachWeek of this.allGames){
			// and get each games from each week
			for (let specificTeamGame of eachWeek){
				if (specificTeamGame["homeTeamRank"] === aRank
						|| specificTeamGame["awayTeamRank"] === aRank){
					// converting time into string, readable time format
					let newDate = new Date(specificTeamGame.dateTime).toDateString()
					specificTeam += newDate + `${View.SPACE()}`
					// getting time like 7:35PM
					// was gonna use navigator.language instead of en-US but it comes up with korean
					// even thought default language is ENG
					let newTime = new Date(specificTeamGame.dateTime).toLocaleTimeString("en-US", {hour: "numeric", minute: "numeric"})

					specificTeam += newTime + `${View.SPACE()}`

					specificTeam += "Week:" + specificTeamGame.week
													+ `${View.SPACE()}`
													// getting hometeam name by rank
													+ "Team: " + this.allTeams[specificTeamGame.homeTeamRank - 1].name
													+ `${View.SPACE()}`
													+ "v"
													+ `${View.SPACE()}`
													// getting awayteam name by rank
													+ this.allTeams[specificTeamGame.awayTeamRank - 1].name
													+ `${View.SPACE()}`
													// getting hometeam venue by rank
													+ "At: " + this.allTeams[specificTeamGame.homeTeamRank -1].venue
													+ `${View.SPACE()}`
													// getting hometeam city by rank
													+ this.allTeams[specificTeamGame.homeTeamRank -1].city
					 								+ `${View.NEWLINE()}`
				}
			}
		}
		return specificTeam
	}

	// getCrossOverGames method, to print out games thats between Premiership and
	// Championship
	getCrossOverGames() {
		let crossovergames = "Will display crossover games" + `${View.NEWLINE()}`
		// allGames will have multiple lists inside as there are more than one week of games
		for (let eachWeek of this.allGames){
			// and get each games from each week
			for (let teamGame of eachWeek){
				// as crossover is between Championship and Premiership, so we have to check
				// both home and away.
				// I also made a list called allPDRanks, allCDRanks so I can use includes
				// builtin function to find whether this list contains the rank we are
				// looking
				if ((this.allPDRanks.includes(teamGame["homeTeamRank"]) &&
							this.allCDRanks.includes(teamGame["awayTeamRank"]))
						||
						(this.allCDRanks.includes(teamGame["homeTeamRank"]) &&
									this.allPDRanks.includes(teamGame["awayTeamRank"]))){
					// again, change time to readable, understandable time format
					let newDate = new Date(teamGame.dateTime).toDateString()
					crossovergames += newDate + `${View.SPACE()}`
					// getting time like 7:35PM
					// was gonna use navigator.language instead of en-US but it comes up with korean
					// even thought default language is ENG
					let newTime = new Date(teamGame.dateTime).toLocaleTimeString("en-US", {hour: "numeric", minute: "numeric"})

					crossovergames += newTime + `${View.SPACE()}`

					crossovergames += "Week:" + teamGame.week
													+ `${View.SPACE()}`
													// unlike other method, its comparing between Premiership and Championship
													// I decided to print out their rank so you can see clearly that those outputs are crossover games.
													+ "Team: rank:" + teamGame.homeTeamRank + `${View.SPACE()}` + this.allTeams[teamGame.homeTeamRank - 1].name
													+ `${View.SPACE()}`
													+ "vs"
													+ `${View.SPACE()}`
													// again, getting away teams rank and their team name.
													+ "rank:" + teamGame.awayTeamRank + `${View.SPACE()}` + this.allTeams[teamGame.awayTeamRank - 1].name
													+ `${View.SPACE()}`
													// getting hometeam's venue
													+ "At: " + this.allTeams[teamGame.homeTeamRank -1].venue
													+ `${View.SPACE()}`
													// getting hometeams' city
													+ this.allTeams[teamGame.homeTeamRank -1].city
					 								+ `${View.NEWLINE()}`
				}
			}
		}
		return crossovergames
	}

	// getAll method, which will be ran to printout on web to see the result.
	getAll() {
		View.out('----------- Teams -----------' + View.NEWLINE())
		View.out(this.getTeams())
		View.out(View.NEWLINE() + '----------- Games -----------' + View.NEWLINE())
		View.out(this.getGames())
		View.out(View.NEWLINE() + '----------- Canterbury Geams -----------' + View.NEWLINE())
		View.out(this.getCanterburyGames('Canterbury') + View.NEWLINE())
		View.out('----------- Crossover Games -----------' + View.NEWLINE())
		View.out(this.getCrossOverGames())
	}
}

File: PreviousCode/Python/competition.py
import datetime
from Game import Game
from Team import Team

# Hence this is just python translated version of competition.js, not gonna leave comments.
class Competition:
    def __init__(self, new_title, new_year):
        self.title = new_title
        self.year = new_year
        self.allPDTeams = []
        self.allCDTeams = []
        self.allTeams = []
        self.allGames = []
        self.allPDRanks = []
        self.allCDRanks = []

    def addTeam(self, newRank, newName, newVenue, newCity):
        newTeam = Team(newRank, newName, newVenue, newCity)
        self.allTeams.append(newTeam)
        if (newRank <= 7):
            self.allPDTeams.append(newTeam)
            self.allPDRanks.append(newTeam.rank)
        else:
            self.allCDTeams.append(newTeam)
            self.allCDRanks.append(newTeam.rank)

    def getTeams(self):
        result = f"TEAMS\n"
        result += f"Premiership Division\n"
        for team in self.allPDTeams:
            result += str(team) + "\n"
        result += f"Championship Division\n"
        for team in self.allCDTeams:
            result += str(team) + "\n"
        return result

    def addGame(self, newWeek, newHomeTeamRank, newAwayTeamRank, newDateTime):
        newGame = Game(newWeek, newHomeTeamRank, newAwayTeamRank, newDateTime)
        self.allGames.append(newGame)

    def getGames(self):
        schedule = str(self.year) + ' ' + self.title + ' Draw\n'
        for aWeek in self.allGames:
            date_time_obj = datetime.datetime.strptime(aWeek.dateTime, '%Y-%m-%dT%H:%M:%S.%fZ')
            schedule += 'Week:' + str(aWeek.week) + '\n'
            schedule += str(date_time_obj.date().strftime("%a %d")) + ', '
            schedule += str(date_time_obj.time().strftime("%H:%MPM")) + ' '
            schedule += 'Team: ' + self.allTeams[aWeek.homeTeamRank - 1].name
            schedule += ' vs ' + self.allTeams[aWeek.awayTeamRank - 1].name
            schedule += ' At: ' + self.allTeams[aWeek.homeTeamRank - 1].venue
            schedule += ', ' + self.allTeams[aWeek.homeTeamRank - 1].city + '\n'

        return schedule

    def getCanterburyGames(self, teamName):
        rank = 0
        specificTeamGames = 'Will only display ' + teamName + " team's games\n"
        for aTeam in self.allTeams:
            if (aTeam.name == teamName):
                rank = aTeam.rank
        for thatTeam in self.allGames:
            if (thatTeam.homeTeamRank == rank or thatTeam.awayTeamRank == rank):
                date_time_obj = datetime.datetime.strptime(thatTeam.dateTime, '%Y-%m-%dT%H:%M:%S.%fZ')
                specificTeamGames += 'Week:' + str(thatTeam.week) + '\n'
                specificTeamGames += str(date_time_obj.date().strftime("%a %d")) + ', '
                # I could've used strftime("%H:%M%p") but this case will allocate as AM as time is 7 not 19 for 7PM.
                specificTeamGames += str(date_time_obj.time().strftime("%H:%MPM")) + ', '
                specificTeamGames += 'Team: ' + self.allTeams[thatTeam.homeTeamRank - 1].name
                specificTeamGames += ' vs ' + self.allTeams[thatTeam.awayTeamRank - 1].name
                specificTeamGames += ' At: ' + self.allTeams[thatTeam.homeTeamRank - 1].venue
                specificTeamGames += ', ' + self.allTeams[thatTeam.homeTeamRank - 1].city + '\n'

        return specificTeamGames

    def getCrossOverGames(self):
        crossovergames = "Will display crossover games\n"

        for eachGame in self.allGames:
            if ((eachGame.homeTeamRank in self.allPDRanks) and (eachGame.awayTeamRank in self.allCDRanks)) or ((eachGame.homeTeamRank in self.allCDRanks) and (eachGame.awayTeamRank in self.allPDRanks)):

                date_time_obj = datetime.datetime.strptime(eachGame.dateTime, '%Y-%m-%dT%I:%M:%S.%fZ')
                crossovergames += 'Week:' + str(eachGame.week) + '\n'
                crossovergames += str(date_time_obj.date().strftime("%a %d")) + ', '
                crossovergames += str(date_time_obj.time().strftime("%H:%MPM")) + ', '
                crossovergames += 'Team: rank:' + str(self.allTeams[eachGame.homeTeamRank - 1].rank) + ' ' + self.allTeams[eachGame.homeTeamRank - 1].name
                crossovergames += ' vs rank:' + str(self.allTeams[eachGame.awayTeamRank - 1].rank) + ' ' + self.allTeams[eachGame.awayTeamRank - 1].name
                crossovergames += ' At: ' + self.allTeams[eachGame.homeTeamRank - 1].venue
                crossovergames += ', ' + self.allTeams[eachGame.homeTeamRank - 1].city + '\n'

        return crossovergames


    def getAll(self):
        result = ''
        result += '----------- Teams -----------\n'
        result += self.getTeams()
        result += '----------- Games -----------\n'
        result += self.getGames()
        result += '----------- Canterbury Games -----------\n'
        result += self.getCanterburyGames('Canterbury')
        result += '----------- Crossover Games -----------\n'
        result += self.getCrossOverGames()

        return result

    def __str__(self):
        return str(self.getAll())
